Added SignTool for v5

jhromadik opened the pull request:
Added SignTool for FAKE v5, some documentation, unit tests and updated obsolete information for the old SignToolHelper.

If someone could please have a look to make sure everything is OK ... Other than that, I think it's ready.

matthid commented:
Thanks @jhromadik!

Can you please resolve the conflicts?

One thing I already noticed: Usually API design is to let users input a function and change arguments. You let users call `Options.Create` and then modify the record. Any particular reason to not follow what other modules do?

jhromadik commented:
Conflicts resolved.

As to my API design choice ... I'm not sure. I thought that I had looked at other projects to see what they were doing, concluded that it was about 50/50, and decided to go with the record-style. Looking through the projects now, I see that everyone is doing the function-style. So now I don't know what I was thinking.

Anyway, I reworked all functions to the function-style by default. I also kept the record-style functions as alternatives, same name only with an 'o' suffix, should anyone prefer to call them that way.

matthid commented:
One problem with the regular record is when we add fields which we don't consider a breaking change. If people construct the record by hand (by specifying all fields and not using the create function) the code will no longer compile. So usung the functions kind of forces people to use the correct syntax without them having to know about the create function. 
I'm not saying that we cannot provide those overloads I'm probably just not yet fully convinced :)

jhromadik commented:
Only function-style functions remain. They have better usability, and if you have a record you want to use, you can still do that - took me a while to realize this :). Added random string generation to tests, since FsCheck doesn't do that by default.

matthid commented on src/app/Fake.Tools.SignTool/SignTool.fs:
```diff
@@ -0,0 +1,403 @@
+        let serverUrl = t.ServerUrl |> Option.defaultValue "http://timestamp.digicert.com"
+        yield match t.Algorithm with
+              | Some SHA1 -> sprintf "/t \"%s\"" serverUrl
+              | Some SHA256 | None -> sprintf "/tr \"%s\" /td \"sha256\"" serverUrl
```
I'd prefer to use the Arguments API as (briefly) documented in https://fake.build/core-process.html and https://fake.build/apidocs/v5/fake-core-argumentsmodule.html
It handles proper escaping and edge cases accordingly

jhromadik replied:
I'll look into it.

matthid commented on src/app/Fake.Tools.SignTool/SignTool.fs:
```diff
@@ -0,0 +1,403 @@
+    | true, File f ->
+        if f.Password.IsSome then
+            // surround in quotes to lower chances of replacing non-password occurences of password-string
+            TraceSecrets.register "\"<PASSWORD>\"" (sprintf "\"%s\"" f.Password.Value)
```
I guess you mean if a very simple password is used and replaced elsewhere?
This is a known problem, even on build servers which support secret variables (like azure devops)
I'm not sure we need special casing here, this might lead to the password being printed when it is not quoted?

jhromadik replied:
I ran into this when I was using a test certificate where the password was also in the file name (to make sure I didn't forget it) - part of the file name was replaced with the registered replacement which looked pretty weird.
The implementation makes sure it is always quoted and there is a test for it too, so erroneously displaying the password should not be a problem.

jhromadik commented on help/markdown/fake-tools-signtool.md:
```diff
@@ -0,0 +1,301 @@
+// val sign : certificate:SignTool.SignCertificate -> setOptions:(SignTool.SignOptions -> SignTool.SignOptions) -> files:seq<string> -> unit
+SignTool.sign
+    (SignTool.SignCertificate.Store {
+        SignTool.CertificateFromStore.Create() with
```
I was trying to find a way to make this "prettier" but came up empty. I like your suggestion and will implement it, thanks.

jhromadik commented on help/markdown/fake-tools-signtool.md:
```diff
@@ -0,0 +1,301 @@
+SignTool.timeStamp
+    (fun o ->
+        { o with
+            TimeStamp = Some { SignTool.TimeStampOption.Create() with
```
I think I'll just pull `ServerUrl` and `Algorithm` up to the options level.

Default parameters:

    SignTool.timeStamp (fun o -> o)

With additional parameters:

    SignTool.timeStamp
        (fun o -> { o with
                        ServerUrl = ...
                        Algorithm = ... } )

jhromadik commented on src/app/Fake.Tools.SignTool/SignTool.fs:
```diff
@@ -0,0 +1,403 @@
+        ToolPath: string option
+        /// Timeout.
+        /// If not provided, default value is 10 seconds per file.
+        Timeout: TimeSpan option
```
I was just trying to set a reasonable default value. In retrospect, it is unreasonable and more confusing than useful. I'll remove the default value and there will be no timeout when set to `None`.

jhromadik commented on help/markdown/fake-tools-signtool.md:
```diff
@@ -0,0 +1,301 @@
+// set path to signtool.exe - if you want to use a specific version or you don't have Windows SDKs installed
+// by default, an attempt will be made to locate it automatically in 'Program Files (x86)\Windows Kits'
+{ o with
+    ToolOptions = Some { SignTool.ToolOptions.Create() with
```
I'll "inline" these.  It will be much prettier on the surface and should only be a minor implementation inconvenience.

Robin1985 commented:
Hi! Thanks for the contribution, is there anything I can do to help get this merged? I was just about to port this but then saw you already had a pull request for it.

matthid commented:
I hope to find the time to prepare a new release within a couple of days. Either all points are addressed (or only small things which I can do myself are open) and it will be part of the release or new comments will follow :)

jhromadik commented:
I think I've addressed all comments:

- added `CreateFromFile` and `CreateFromStore` static helpers
- made signing with timestamp a separate function `singWithTimeStamp` with additional parameters
- no `Timeout` means no timeout
- "inlined" `ToolOptions`
- revorked implementation to use Arguments API

Plus some additional work:

- removed all defaults that were different from signtool.exe's defaults
- improved test to make sure final arguments only contain exactly what they are supposed to contain

@matthid, @Robin1985 please have a look at it and let me know what you think.

matthid commented:
Looks good to me, thanks a lot!
